notebook/notebookapp.py

```diff
@@ -704,6 +708,43 @@ class NotebookApp(JupyterApp):
 
     _log_formatter_cls = LogFormatter
 
+    _json_logging_import_error_logged = False
+
+    log_json = Bool(False, config=True,
+        help=_('Set to True to enable JSON formatted logs. '
+               'Run "pip install notebook[json-logging]" to install the '
+               'required dependent packages. Can also be set using the '
+               'environment variable JUPYTER_ENABLE_JSON_LOGGING=true.')
+    )
+
+    @default('log_json')
+    def _default_log_json(self):
+        """Get the log_json value from the environment."""
+        return os.getenv('JUPYTER_ENABLE_JSON_LOGGING', 'false').lower() == 'true'
+
+    @validate('log_json')
+    def _validate_log_json(self, proposal):
+        # If log_json=True, see if the json_logging package can be imported and
+        # override _log_formatter_cls if so.
+        value = proposal['value']
+        if value:
+            try:
+                import json_logging
+                self.log.debug('initializing json logging')
```

Comment:
I'm finding this is never produced and am thinking it should be removed.

Reply:
How are you running things? Because I see it like this:

```console
(.tox) osboxes@osboxes:~/jupyter/notebook$ JUPYTER_ENABLE_JSON_LOGGING=true jupyter notebook --debug
[D 15:25:58.460 NotebookApp] Searching ['/home/osboxes/jupyter/notebook', '/home/osboxes/.jupyter', '/home/osboxes/jupyter/notebook/.tox/etc/jupyter', '/usr/local/etc/jupyter', '/etc/jupyter'] for config files
[D 15:25:58.461 NotebookApp] Looking for jupyter_config in /etc/jupyter
[D 15:25:58.461 NotebookApp] Looking for jupyter_config in /usr/local/etc/jupyter
[D 15:25:58.461 NotebookApp] Looking for jupyter_config in /home/osboxes/jupyter/notebook/.tox/etc/jupyter
[D 15:25:58.461 NotebookApp] Looking for jupyter_config in /home/osboxes/.jupyter
[D 15:25:58.461 NotebookApp] Looking for jupyter_config in /home/osboxes/jupyter/notebook
[D 15:25:58.461 NotebookApp] Looking for jupyter_notebook_config in /etc/jupyter
[D 15:25:58.461 NotebookApp] Looking for jupyter_notebook_config in /usr/local/etc/jupyter
[D 15:25:58.462 NotebookApp] Looking for jupyter_notebook_config in /home/osboxes/jupyter/notebook/.tox/etc/jupyter
[D 15:25:58.462 NotebookApp] Looking for jupyter_notebook_config in /home/osboxes/.jupyter
[D 15:25:58.462 NotebookApp] Looking for jupyter_notebook_config in /home/osboxes/jupyter/notebook
[D 15:25:58.465 NotebookApp] Paths used for configuration of jupyter_notebook_config: 
    	/etc/jupyter/jupyter_notebook_config.json
[D 15:25:58.465 NotebookApp] Paths used for configuration of jupyter_notebook_config: 
    	/usr/local/etc/jupyter/jupyter_notebook_config.json
[D 15:25:58.466 NotebookApp] Paths used for configuration of jupyter_notebook_config: 
    	/home/osboxes/jupyter/notebook/.tox/etc/jupyter/jupyter_notebook_config.json
[D 15:25:58.466 NotebookApp] Paths used for configuration of jupyter_notebook_config: 
    	/home/osboxes/.jupyter/jupyter_notebook_config.json
[D 15:25:58.476 NotebookApp] initializing json logging
{"written_at": "2020-10-12T20:25:58.650Z", "written_ts": 1602534358650364000, "msg": "Serving notebooks from local directory: /home/osboxes/jupyter/notebook", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "INFO", "module": "notebookapp", "line_no": 2250}
{"written_at": "2020-10-12T20:25:58.650Z", "written_ts": 1602534358650523000, "msg": "Jupyter Notebook 7.0.0.dev0 is running at:", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "INFO", "module": "notebookapp", "line_no": 2250}
...
```

Reply:
Assuming you have `json-logging` installed too...`pip install -e .[json-logging]`. Otherwise you'd hit the ImportError and skip that debug line.

Reply:
I see the expected behavior when the package is not installed, etc.  but it does appear there's a difference when the env is used vs. the CLI option.  In the latter case, here are the results, up to the same period you list above.  All of the logging is JSON but I don't see the particular debug message produced in the validation handler - thus I figured it's not useful.  Since it appears in some cases, I suppose it does provide some value - especially since in this hybrid log format case, it would be helpful to know the sudden JSON output is intentional.  My hunch is that you'd see the same mixed results when configured via a config file, whereas CLI options take precedence and, I suppose, are processed prior to the others as things get loaded.

```
$ jupyter notebook --debug --notebook-dir=~/notebooks --NotebookApp.log_json=True
{"written_at": "2020-10-12T21:17:02.042Z", "written_ts": 1602537422042635000, "msg": "Searching ['/Users/kbates/repos/oss/jupyter/notebook', '/Users/kbates/.jupyter', '/opt/anaconda3/envs/notebook-dev/etc/jupyter', '/usr/local/etc/jupyter', '/etc/jupyter'] for config files", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "application", "line_no": 179}
{"written_at": "2020-10-12T21:17:02.042Z", "written_ts": 1602537422042879000, "msg": "Looking for jupyter_config in /etc/jupyter", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "application", "line_no": 556}
{"written_at": "2020-10-12T21:17:02.043Z", "written_ts": 1602537422043005000, "msg": "Looking for jupyter_config in /usr/local/etc/jupyter", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "application", "line_no": 556}
{"written_at": "2020-10-12T21:17:02.043Z", "written_ts": 1602537422043118000, "msg": "Looking for jupyter_config in /opt/anaconda3/envs/notebook-dev/etc/jupyter", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "application", "line_no": 556}
{"written_at": "2020-10-12T21:17:02.043Z", "written_ts": 1602537422043214000, "msg": "Looking for jupyter_config in /Users/kbates/.jupyter", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "application", "line_no": 556}
{"written_at": "2020-10-12T21:17:02.043Z", "written_ts": 1602537422043305000, "msg": "Looking for jupyter_config in /Users/kbates/repos/oss/jupyter/notebook", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "application", "line_no": 556}
{"written_at": "2020-10-12T21:17:02.043Z", "written_ts": 1602537422043996000, "msg": "Looking for jupyter_notebook_config in /etc/jupyter", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "application", "line_no": 556}
{"written_at": "2020-10-12T21:17:02.044Z", "written_ts": 1602537422044112000, "msg": "Looking for jupyter_notebook_config in /usr/local/etc/jupyter", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "application", "line_no": 556}
{"written_at": "2020-10-12T21:17:02.044Z", "written_ts": 1602537422044218000, "msg": "Looking for jupyter_notebook_config in /opt/anaconda3/envs/notebook-dev/etc/jupyter", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "application", "line_no": 556}
{"written_at": "2020-10-12T21:17:02.044Z", "written_ts": 1602537422044313000, "msg": "Looking for jupyter_notebook_config in /Users/kbates/.jupyter", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "application", "line_no": 556}
{"written_at": "2020-10-12T21:17:02.044Z", "written_ts": 1602537422044774000, "msg": "Loaded config file: /Users/kbates/.jupyter/jupyter_notebook_config.py", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "application", "line_no": 578}
{"written_at": "2020-10-12T21:17:02.044Z", "written_ts": 1602537422044951000, "msg": "Looking for jupyter_notebook_config in /Users/kbates/repos/oss/jupyter/notebook", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "application", "line_no": 556}
{"written_at": "2020-10-12T21:17:02.045Z", "written_ts": 1602537422045964000, "msg": "Raising open file limit: soft 2560->4096; hard 9223372036854775807->9223372036854775807", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "notebookapp", "line_no": 1661}
{"written_at": "2020-10-12T21:17:02.049Z", "written_ts": 1602537422049537000, "msg": "Paths used for configuration of jupyter_notebook_config: __/etc/jupyter/jupyter_notebook_config.json", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "config_manager", "line_no": 97}
{"written_at": "2020-10-12T21:17:02.049Z", "written_ts": 1602537422049934000, "msg": "Paths used for configuration of jupyter_notebook_config: __/usr/local/etc/jupyter/jupyter_notebook_config.json", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "config_manager", "line_no": 97}
{"written_at": "2020-10-12T21:17:02.050Z", "written_ts": 1602537422050385000, "msg": "Paths used for configuration of jupyter_notebook_config: __/opt/anaconda3/envs/notebook-dev/etc/jupyter/jupyter_notebook_config.d/elyra.json__/opt/anaconda3/envs/notebook-dev/etc/jupyter/jupyter_notebook_config.d/jupyterlab_git.json__/opt/anaconda3/envs/notebook-dev/etc/jupyter/jupyter_notebook_config.d/nbdime.json__/opt/anaconda3/envs/notebook-dev/etc/jupyter/jupyter_notebook_config.json", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "config_manager", "line_no": 97}
{"written_at": "2020-10-12T21:17:02.051Z", "written_ts": 1602537422051203000, "msg": "Paths used for configuration of jupyter_notebook_config: __/Users/kbates/.jupyter/jupyter_notebook_config.json", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "config_manager", "line_no": 97}
{"written_at": "2020-10-12T21:17:03.096Z", "written_ts": 1602537423096166000, "msg": "Using default logger", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "loader", "line_no": 342}
{"written_at": "2020-10-12T21:17:03.096Z", "written_ts": 1602537423096397000, "msg": "Using default logger", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "loader", "line_no": 342}
{"written_at": "2020-10-12T21:17:03.096Z", "written_ts": 1602537423096496000, "msg": "Using default logger", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "loader", "line_no": 342}
{"written_at": "2020-10-12T21:17:03.096Z", "written_ts": 1602537423096575000, "msg": "Using default logger", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "loader", "line_no": 342}
{"written_at": "2020-10-12T21:17:03.096Z", "written_ts": 1602537423096649000, "msg": "Using default logger", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "DEBUG", "module": "loader", "line_no": 342}
{"written_at": "2020-10-12T21:17:03.098Z", "written_ts": 1602537423098998000, "msg": "Serving notebooks from local directory: /Users/kbates/notebooks", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "INFO", "module": "notebookapp", "line_no": 2250}
{"written_at": "2020-10-12T21:17:03.099Z", "written_ts": 1602537423099071000, "msg": "Jupyter Notebook 7.0.0.dev0 is running at:", "type": "log", "logger": "NotebookApp", "thread": "MainThread", "level": "INFO", "module": "notebookapp", "line_no": 2250}
```